### ISSUE TITLE
Suggestion; idle disconnection options

### ISSUE THREAD

**Issue by khaytsus** [ubergeek42/weechat-android#510]
It would be really handy to have an option that allows the client to automatically disconnect from the weechat server if the client has been idle for a period of time.  Of could one could disable this (and would likely be the default) if they want to stay connected, otherwise for battery savings etc it'd be nice to disconnect automatically if idle.  And perhaps a companion option that if it was last auto-disconnected, auto-reconnect on launch, so that the user could have it not connected to the server all the time but also would not have to connect/disconnect manually.

**Comment by oakkitten** [comment]
In my experience battery drain due to w-a is negligible; is this a problem for you?

**Comment by khaytsus** [comment]
Honestly I had not tested it myself, which I should have, but I've heard a few times that battery usage was significant. They were primarily mobile irc users so to be honest they may not even want to client to ever disconnect automatically. For me using it as a secondary client I would want it to in case I didn't disconnect manually.

**Comment by oakkitten** [comment]
if there's a problem with the battery we can look into that. 

otherwise, i just don't think something like that is a task for w-a. i don't suppose it's possible to start/stop the service (the part of the app that keeps the connection) directly using automation (we could support that... maybe), but it should be trivial to employ Tasker et al to tap on the disconnect button in the notification?

**Comment by luk1337** [comment]
@oakkitten I personally stopped using weechat-android due to battery issues. I'd pretty much have to charge my phone every day as opposed to every other day. I think it's due to me having 6 wee-slack instances plus plenty of alive channels connected to it.

**Comment by oakkitten** [comment]
@luk1337 was weechat-android high in battery usage statistics? currently on my device it sits at 5% with 17 hours of use, just above whatsapp at 4% with 20 minutes

it is also possible that these numbers are misleading. the app keeps the connection awake, and while that doesn't require cpu per se, receiving messages does. if the cpu somehow drains a lot of battery there, this will be bad, and the system might not be registering w-a as the one app at fault. when the cpu wakes up other apps will jump at the opportunity an do stuff as well

i tried benchmarking by not touching my phone while keeping w-a connected and it drained about 1-2% overnight, which is the same as with w-a not connected, but this kind of test is unreliable as well due how imprecise battery measuring can be

anyways, i tried some random tasker-ish app from play store and it could easily make w-a disconnect. so this could be a simple workaround for now

**Comment by luk1337** [comment]
>was weechat-android high in battery usage statistics?
No, it wasn't thus it took me long time to figure out what was causing my battery drain...

**Comment by oakkitten** [comment]
aha. well this is unfortunate. if you are willing to benchmark it, it would help to know the difference in battery drain between connected & not connected w-a over a long period of time. also it would be helpful to benchmark w-a with the option "only sync open buffers" on (and sufficiently big ping idle time, say, 10 minutes). if you turn on that option, and close all buffers, there will be no communication between w-a and weechat save for a hotlist synchronization every 5 minutes.

also, what's your device & android version?

**Comment by luk1337** [comment]
I have a lot of devices but let's say I'll try this with OnePlus 8 Pro running LineageOS 18.1 (Android 11; no gapps).

**Comment by oakkitten** [comment]
aha. any difference battery usage-wise between 18.1 and 17.1 that you know of? i'll also try benchmarking (again) on my Xiaomi Mi A2 (LineageOS 17.1, Android 10, gapps)

**Comment by luk1337** [comment]
I don't really know. I stopped using it on Android 10 afaict.

**Comment by luk1337** [comment]
Tested disconnected/connected with default settings:

### Disconnected
* April 3, 21:01 - 100%
* April 4, 10:17 - 97%
### Connected (default settings)
* April 4, 10:35 - 100%
* April 5, 00:09 - 93%

Now charging my phone and trying only sync open buffers and only sync open buffers + ping idle 600 seconds next.

**Comment by luk1337** [comment]
Hmm, "only sync open buffers" seems to work fine now. Only got a 2% drain over 12 hours, maybe because I only had 3 IRC channels open... I'll try opening few slack channels just to see if that changes anything.

**Comment by luk1337** [comment]
As soon as I opened +3 slack channels I got it to drain 2% in 5h so idk, maybe wee-slack is keeping relay socket more alive than it should be.

**Comment by oakkitten** [comment]
i keep forgetting to do my own benchmark sorry (but i will get there)

so it's 3% drain without w-a vs 7% with w-a in approximately the same time, or about 2x drain. i did mention 1-2% drain earlier; i guess i worded that poorly but it was in addition to regular drain of about the same 1-2%, so 1-2% without w-a and 2-4% with w-a. in other words the result was about the same, except i had a better battery

(i'm not sure if these numbers are very useful at large as you can usually expect an error of some 5-20% depending on the device, its age, usage pattern, temperature, etc)

i would say that a battery drain of 4% in 13 hours is not a huge problem though?

**Comment by luk1337** [comment]
>i would say that a battery drain of 4% in 13 hours is not a huge problem though?
I think it would maybe behave worse on a device that actually has any apps installed etc as I clearly remember having to charge my 7 Pro at a time every day vs every other day without weechat.

**Comment by oakkitten** [comment]
yeah as i said other apps might jump at the opportunity and do stuff... perhaps it was just one rogue app? or maybe not. or maybe we had bugs that got fixed? i guess i better run the benchmark on my everyday use device

**Comment by oakkitten** [comment]
so uhm (mind that Mi A2 is absolutely not renowned for its great battery life but still)

* with weechat-android in full sync (~40 buffers): 65% to 57% in 8 hours
* with weechat-android disconnected: 50% to 49% in 6 hours

so maybe we do need to address this after all?

i don't like the idea of disconnecting completely, but maybe we could do something like this:

* after a set time of inactivity, desynchronize all buffers
  * alternatively, desynchronize buffers individually
* after all buffers desynchronized, gradually reduce hotlist sync frequency
* accordingly increase ping idle time (it's independent from the above)
* when the app is active, resume synchronizing

also,

* keep everything in sync if not on battery

the biggest problem here is hot messages, we can't just sync only them

**Comment by khaytsus** [comment]
@oakkitten Could also of course make it an option so that if a user is okay not getting highlights and prefers to save battery they can enable the battery saving options.  Could possibly provide a few levels of savings that immediately disconnects, slowly disconnects as you're thinking, never disconnects, etc.

**Comment by oakkitten** [comment]
so i've been toying with the code and this seems to work? kind of? it would take a bit of time figuring out if there's any race conditions in the logic but the part that syncs/desyncs (also the battery part) seems to work fine!

i think this could be a generic solution to replace the "optimize traffic" option entirely with more options, namely:

* global desync delay — this would say how long app should wait before desyncing all buffers. conveniently it simply calls `desync`. all buffers are synced when app becomes active or the device gets plugged in, and the countdown starts when it becomes inactive.
* desync delay for buffers — amount of time to wait since last visited a buffer. should probably be larger than above ↑
* desync delay for open buffers — same, but for open buffers. not sure if needed but can mimic behavior of "only sync open buffers" if someone is relying on it

given delays of 5, 30, and 60 minutes, accordingly, it can go like this...

* T+0: you open app on an unplugged device 
   you open buffers A and B, close buffer B, minimize app   
   all buffers synchronized
* T+5 global desync. only buffers A and B remain synchronized
* T+30 buffer B is desynced. only buffer A is synced
* T+60 buffer A is desynced. no buffers are synced
* T+90 you plug in device. global sync. all buffers synced
* T+95 you unplug device. since app became inactive a long time ago, all buffers are immediately desynced

**Comment by oakkitten** [comment]
> * accordingly increase ping idle time (it's independent from the above)

actually tying ping idle time to hotlist sync sounds good? if you ping more frequently, why not sync hotlist as resource-wise it's about the same thing; if you ping less frequently, in the normal course of events you'll be never sending pings anyway. (we are not looking at latency atm)

> * after all buffers desynchronized, gradually reduce hotlist sync frequency

what would be a good way of doing this? i'm thinking of something like... (currently it's hardcoded as once 5 minutes)

* if app is open, sync every ½ minute
* if not but synced globally, sync every minute (as we are still expecting high-ish traffic here)
* if not synced globally, but some buffers remain synced, sync once 5 minutes (we are expecting *some* traffic)
* if not syncing any buffers, sync once every... 15 minutes sounds ok? or maybe reduce linearly over a course of a couple of hours or so

should this be configurable in some way? also, this would mean that in the worst case scenario this would mean that the app will not reconnect for half an our after losing connection
